src/common/fileUpload/FileUpload.tsx

```diff
@@ -0,0 +1,205 @@
+type PersistedFileContainer = {
+  name: string;
+  id?: string;
+  data?: never;
+  markedForDeletion?: boolean;
+};
+
+type NewFileContainer = {
+  name: string;
+  id?: never;
+  data?: File;
+  markedForDeletion?: never;
+};
+
+export type FileContainer = PersistedFileContainer | NewFileContainer;
+
+export interface FileUploadProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'value' | 'onChange'> {
+  buttonProps?: FileUploadButtonProps;
+  disabled?: boolean;
+  helperText?: string;
+  invalid?: boolean;
+  label?: string;
+  /* max size in bytes */
+  maxSize?: number;
+  multiple?: boolean;
+  name: string;
+  onChange: (value: undefined | FileContainer | FileContainer[]) => void;
+  value: undefined | FileContainer | FileContainer[];
+}
+
+const FileUpload: FunctionComponent<FileUploadProps> = ({
+  buttonProps = {
+    size: 'small',
+    color: 'supplementary',
+  },
+  disabled,
+  helperText,
+  invalid = false,
+  label,
+  maxSize,
+  multiple = false,
+  name,
+  onChange,
+  value,
+  ...rest
+}) => {
+  const { t, i18n } = useTranslation();
+
+  const handleChange: React.InputHTMLAttributes<HTMLInputElement>['onChange'] = (event) => {
+    if (event.currentTarget.files) {
+      const eventFiles: FileContainer[] = Array.from(event.currentTarget.files).map((file) => ({
+        name: file.name,
+        data: file,
+      }));
+
+      if (multiple) {
+        return onChange(value === undefined ? eventFiles : (value as FileContainer[]).concat(eventFiles));
+      }
+
+      // single
+      return onChange(eventFiles[0]);
+    }
+  };
+
+  const handleDelete = (targetFile: FileContainer) => {
+    // multiple
+    if (Array.isArray(value)) {
+      return onChange(
+        value.reduce((acc: FileContainer[], file: FileContainer) => {
+          if (file === targetFile) {
+            if (file.data === undefined) {
+              // Persisted file, mark for deletion
+              return acc.concat({
+                ...(file as PersistedFileContainer),
+                markedForDeletion: !file.markedForDeletion,
+              });
+            } else {
+              // New file, remove reference
+              return acc;
+            }
+          } else {
+            return acc.concat(file);
+          }
+        }, [])
+      );
+    }
+
+    // single
+    return onChange(
+      (value as FileContainer).data === undefined
+        ? // Persisted file, mark for deletion
+          {
+            ...(value as PersistedFileContainer),
+            markedForDeletion: !(value as PersistedFileContainer).markedForDeletion,
+          }
+        : // New file, remove reference
+          undefined
+    );
+  };
+
+  const renderFileList = () => {
+    if (value === undefined) return undefined;
+
+    const valueList = Array.isArray(value) ? value : [value];
+    return (
+      <List noBullets>
+        {valueList.map((file) => {
+          return (
+            <li key={file.name} className={styles.fileListItem}>
```

Comment:
This was a great point, I added `uuid` to FileContainers to identify them

Reply:
That works; Or simpler but not an ideal solution is using the elements' indexes since there will be only a handful of files.
```TSX
{valueList.map((file, i) => {
          return (
            <li key={i} className={styles.fileListItem}>
```